rust-symcrypt/src/gcm.rs

```diff
@@ -0,0 +1,377 @@
+use core::ffi::c_void;
+use std::mem;
+use std::ptr;
+use std::vec;
+use symcrypt_sys;
+use crate::errors::SymCryptError;
+
+pub struct GcmState {
+    expanded_key: symcrypt_sys::SYMCRYPT_GCM_EXPANDED_KEY,
+    state: symcrypt_sys::SYMCRYPT_GCM_STATE,
+}
+
+impl GcmState {
+    pub fn new(key: &[u8], nonce: &[u8]) -> Result<Box<Self>, SymCryptError> {
+        let mut instance = Box::new(GcmState {
+            state: symcrypt_sys::SYMCRYPT_GCM_STATE::default(),
+            expanded_key: symcrypt_sys::SYMCRYPT_GCM_EXPANDED_KEY::default()
+        });
+        unsafe {
+            expand_key(key, &mut instance.expanded_key)?;
+            symcrypt_sys::SymCryptGcmInit(
+                &mut instance.state,
+                &instance.expanded_key,
+                nonce.as_ptr(),
+                nonce.len() as symcrypt_sys::SIZE_T,
+            );
+            Ok(instance)
+        }
+    }
+
+    pub fn auth(&mut self, auth_data: &[u8]) {
+        unsafe {
+            symcrypt_sys::SymCryptGcmAuthPart(
+                &mut self.state,
+                auth_data.as_ptr(),
+                auth_data.len() as symcrypt_sys::SIZE_T,
+            );
+        }
+    }
+
+    pub fn encrypt_part(&mut self, data: &[u8]) -> Vec<u8> {
+        unsafe {
+            let mut dst: Vec<u8> = vec![0u8; data.len()];
+            symcrypt_sys::SymCryptGcmEncryptPart(
+                &mut self.state,
+                data.as_ptr(),
+                dst.as_mut_ptr(),
+                dst.len() as symcrypt_sys::SIZE_T,
+            );
+            dst
+        }
+    }
+
+    pub fn encrypt_final(&mut self, tag: &mut [u8]) {
```

Comment:
Size of `tag` is not known at compile time, taking it as an out parameter so caller can specify. The alternative is to take in a tag length. I'm okay with doing either, having the out parameter seems better since it makes both Encrypt/Decrypt have the same call signature

Reply:
After further thinking, recommend tag_length parameter and return tag in the return result.

Reply:
What changed your mind?

Reply:
So far we haven't had any out parameters. Only out via the Result. Seems better to be consistent throughout.

Reply:
One advantage of the out parameter is that it lets the caller control memory allocation, while if we take a length and return the tag, that means `encrypt_final` will have to handle memory allocation (probably by returning a `Vec<u8>`). Some applications are very sensitive to memory allocation, so that extra control can be important, but it depends on the requirements you expect your users to have.

Reply:
Going to go with returning a tuple, I think it is more clear and the trade off of memory allocation is just a sacrifice we'll make.